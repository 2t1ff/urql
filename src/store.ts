import { Entity, EntityMap, KeyExtractor, Link, LinkMap } from './types';
import {
  assignObjectToMap,
  isOperation,
  keyOfEntity,
  objectOfMap,
} from './utils';

export interface StoreData {
  records: {
    [key: string]: Entity;
  };
  links: {
    [key: string]: Link;
  };
}

export interface StoreOpts {
  initial?: StoreData;
  keyExtractor?: KeyExtractor;
}

class Store {
  private touched: string[];
  private records: EntityMap;
  private links: LinkMap;

  public keyOfEntity: (entity: Entity) => null | string;

  constructor({ initial, keyExtractor }: StoreOpts) {
    this.touched = [];
    this.records = new Map();
    this.links = new Map();

    if (keyExtractor !== undefined) {
      this.keyOfEntity = (entity: Entity) => {
        const key = keyExtractor(entity);
        return key !== undefined ? key : keyOfEntity(entity);
      };
    } else {
      this.keyOfEntity = keyOfEntity;
    }

    if (initial !== undefined) {
      assignObjectToMap(this.records, initial.records);
      assignObjectToMap(this.links, initial.links);
    }
  }

  getEntity(key: string): null | Entity {
    if (!isOperation(key)) {
      this.touched.push(key);
    }

    const entity = this.records.get(key);
    return entity !== undefined ? entity : null;
  }

  getOrCreateEntity(key: string): Entity {
    const prev = this.records.get(key);
    if (prev !== undefined && prev !== null) {
      return prev;
    }

    const entity = Object.create(null);
    this.records.set(key, entity);
    return entity;
  }

  writeEntityValue(key: string, prop: string, val: any) {
    if (!isOperation(key)) {
      this.touched.push(key);
    }

    const entity = this.getOrCreateEntity(key);
    if (val === undefined) {
      delete entity[prop];
    } else {
      entity[prop] = val;
    }
  }

  getLink(key: string): Link {
    this.touched.push(key);
    const link = this.links.get(key);
    return link !== undefined ? link : null;
  }

  writeLink(key: string, link: Link) {
    this.touched.push(key);

    if (link === null) {
      this.links.delete(key);
    } else {
      this.links.set(key, link);
    }
  }

  toJSON(): StoreData {
    return {
      records: objectOfMap(this.records),
      links: objectOfMap(this.links),
    };
  }

  flushTouched(): string[] {
    const touched = this.touched.filter((key, i, arr) => {
      return arr.indexOf(key) === i;
    });

    this.touched = [];
    return touched;
  }
}

export default Store;
