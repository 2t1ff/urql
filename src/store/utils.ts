export const assignObjectToMap = <T>(
  map: Map<string, T>,
  obj: { [key: string]: T }
) => {
  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      map.set(key, obj[key]);
    }
  }
};

export const objectOfMap = <T>(map: Map<string, T>): { [key: string]: T } => {
  const res = {};
  map.forEach((value, key) => {
    res[key] = value;
  });
  return res;
};
