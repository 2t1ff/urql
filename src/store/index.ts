import { Entity, Link, Links, Records } from '../types';
import { assignObjectToMap, objectOfMap } from './utils';

export interface Store {
  records: Records;
  links: Links;
}

export interface SerializedStore {
  records: { [link: string]: Entity };
  links: { [link: string]: Link };
}

/** Creates a new Store with an optional initial, serialized state */
export const create = (initial?: SerializedStore): Store => {
  const records: Records = new Map();
  const links: Links = new Map();

  if (initial !== undefined) {
    assignObjectToMap(records, initial.records);
    assignObjectToMap(links, initial.links);
  }

  return { records, links };
};

/** Serializes a given Store to a plain JSON structure */
export const serialize = (store: Store): SerializedStore => {
  const records = objectOfMap(store.records);
  const links = objectOfMap(store.links);
  return { records, links };
};

export const find = (store: Store, key: string): Entity | null => {
  const entity = store.records.get(key);
  return entity !== undefined ? entity : null;
};

export const findOrCreate = (store: Store, key: string): Entity => {
  const entity = find(store, key);
  if (entity !== null) {
    return entity;
  }

  const record: Entity = Object.create(null);
  store.records.set(key, record);
  return record;
};

export const setLink = (store: Store, key: string, link: Link): void => {
  store.links.set(key, link);
};

export const deleteLink = (store: Store, key: string): void => {
  store.links.delete(key);
};

export const readLink = (store: Store, key: string): void | Link =>
  store.links.get(key);
