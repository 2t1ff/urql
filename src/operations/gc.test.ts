import gql from 'graphql-tag';
import { create, serialize } from '../store';
import { gc, write } from './index';

const gqlQuery = gql`
  {
    todos {
      __typename
      id
      text
      complete
    }
    todo {
      __typename
      id
      text
      complete
      author {
        __typename
        name
      }
    }
  }
`;

it('cleans up unreachable records and links', () => {
  const store = create();

  write(
    store,
    { query: gqlQuery },
    {
      todos: [
        {
          id: '0',
          text: 'Go to the shops',
          complete: false,
          __typename: 'Todo',
        },
        {
          id: '1',
          text: 'Pick up the kids',
          complete: true,
          __typename: 'Todo',
        },
        { id: '2', text: 'Install urql', complete: false, __typename: 'Todo' },
      ],
      todo: {
        __typename: 'Todo',
        id: '2',
        text: 'Install urql',
        complete: false,
        author: {
          __typename: 'Author',
          name: 'Thomas',
        },
      },
    }
  );

  let json = serialize(store);

  expect(json.links['query.todos']).toEqual(['Todo:0', 'Todo:1', 'Todo:2']);
  expect(json.records['Todo:1']).toMatchObject({ __typename: 'Todo' });
  expect(json.records['Todo:2.author']).toMatchObject({ __typename: 'Author' });

  write(
    store,
    { query: gqlQuery },
    {
      todos: null,
      todo: {
        __typename: 'Todo',
        id: '2',
        text: 'Install urql',
        complete: false,
        author: {
          __typename: 'Author',
          name: 'Thomas',
        },
      },
    }
  );

  json = serialize(store);

  // Same as above; everything is still reachable but `query.todos` is gone
  expect(json.links['query.todos']).toBe(undefined);
  expect(json.records['Todo:1']).toMatchObject({ __typename: 'Todo' });
  expect(json.records['Todo:2.author']).toMatchObject({ __typename: 'Author' });

  gc(store);
  json = serialize(store);

  expect(json).toMatchSnapshot();
  expect(json.links['query.todos']).toBe(undefined);
  expect(json.records['Todo:1']).toBe(undefined);
  expect(json.records['Todo:2']).not.toBe(undefined);
  expect(json.records['Todo:2.author']).toMatchObject({ __typename: 'Author' });

  // Remove all results
  write(
    store,
    { query: gqlQuery },
    {
      todos: null,
      todo: null,
    }
  );

  gc(store);
  json = serialize(store);

  expect(json).toEqual({
    records: {
      query: {
        todos: null,
        todo: null,
      },
    },
    links: {},
  });
});
