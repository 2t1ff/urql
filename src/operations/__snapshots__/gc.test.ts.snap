// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleans up unreachable records and links 1`] = `
Object {
  "links": Object {
    "Todo:2.author": "Todo:2.author",
    "query.todo": "Todo:2",
  },
  "records": Object {
    "Todo:2": Object {
      "__typename": "Todo",
      "author": null,
      "complete": false,
      "id": "2",
      "text": "Install urql",
    },
    "Todo:2.author": Object {
      "__typename": "Author",
      "name": "Thomas",
    },
    "query": Object {
      "todo": null,
      "todos": null,
    },
  },
}
`;
